### PR TITLE
Parse fonts

### PR DESCRIPTION
**Pull request by dmorais-ledger**
## Remove OCR for LNX, LNS+ & Stax devices
* Parse ELF data to find fonts information on LNX & LNS+ (for Stax, fonts are loaded at a known location).
* Use those information to build a bitmap -> character array
* No need to change Bolos nor the SDK!
* All active API levels are supported (1, 5 & 12)

NB: legacy OCR is kept for SDK prior to unification.

### CODE REVIEW

File: speculos/main.py [hunk 1]
```diff
@@ -106,11 +106,20 @@ def get_elf_infos(app_path):
...
             svc_cx_call_addr = svc_cx_call_symbol[0]['st_value'] & (~1)
+        # Check where are located fonts in .elf file (LNX/LNS+ only)
+        # (Stax fonts are loaded at a known location: STAX_FONTS_ARRAY_ADDR)
+        fonts_addr = 0
+        fonts_size = 0
+        bagl_fonts_symbol = symtab.get_symbol_by_name('C_bagl_fonts')
+        if bagl_fonts_symbol is not None:
+            fonts_addr = bagl_fonts_symbol[0]['st_value']
+            fonts_size = bagl_fonts_symbol[0]['st_size']
+            print(f"Found C_bagl_fonts at 0x{fonts_addr:X} ({fonts_size} bytes)\n")
```
Comment (xchapron-ledger) [inline speculos/main.py:117]: log.info maybe?

File: src/bolos/nbgl.c [hunk 2]
```diff
@@ -53,15 +48,16 @@ unsigned long sys_nbgl_front_draw_horizontal_line(nbgl_area_t *area,
   return 0;
 }
 
-unsigned long sys_nbgl_front_draw_img(nbgl_area_t *area, uint8_t *buffer,
-                                      nbgl_transformation_t transformation,
-                                      nbgl_color_map_t colorMap)
+static unsigned long sys_nbgl_front_draw_img_character(nbgl_area_t *area, uint8_t *buffer,
```
Comment (xchapron-ledger) [inline src/bolos/nbgl.c:51]: Maybe this should'nt be prefixed with sys as this is not a real syscall?

**Review by xchapron-ledger — APPROVED** [review]

**Review by xchapron-ledger — COMMENTED** [review]
Some CI issues to fix too, else looks promising!

File: src/bolos/bagl.c [hunk 3]
```diff
@@ -92,7 +101,7 @@ unsigned long sys_bagl_hal_draw_bitmap_within_rect(
   size_t i, len, size;
   uint8_t buf[300 - 4]; /* size limit in io_seph_send */
   uint8_t header[4];
-
+  uint32_t character = get_character_from_bitmap(bitmap);
```
Comment (xchapron-ledger) [inline src/bolos/bagl.c:104]: could be skipped on NANOS, if this syscall is really called on this device?

File: speculos/main.py (hunk 1 above)
Comment (dmorais-ledger) [reply speculos/main.py:117]: It was a debug info => I will remove it.

File: speculos/mcu/ocr.py [hunk 4]
```diff
@@ -277,47 +187,44 @@ def store_char_in_last_event(self, x: int, y: int, w: int, h: int, char: str) ->
...
             # In both cases, data contains:
             # - area (sizeof(nbgl_area_t))
             # - compressed bitmap (buffer_len)
             # - 2 bytes of different meaning depending on SephTag
+            # - 4 bytes with unicode code point of displayed character
             area = nbgl_area_t.parse(data[0:nbgl_area_t.sizeof()])
             x, y, w, h = area.x0, area.y0, area.width, area.height
-            bitmap = data[nbgl_area_t.sizeof():-2]
+            character = int.from_bytes(data[-4:], byteorder="little", signed=False)
+            bitmap = data[nbgl_area_t.sizeof():-6]
```
Comment (lpascal-ledger) [inline speculos/mcu/ocr.py:227]:
```suggestion
            bitmap = data[nbgl_area_t.sizeof():-(2+4)]
```

File: src/bolos/bagl.c (hunk 3 above)
Comment (dmorais-ledger) [reply src/bolos/bagl.c:104]: character is supposed to be always present in data, even for unsupported devices/api_levels.
This function will return 0, which is the expected behaviour in that case.

**Review by lpascal-ledger — COMMENTED** [review]
+ some misspells pointed by the CI.

Nice to see such code simplification.

**Review by xchapron-ledger — APPROVED** [review]

**Pull request merged**